File: implementations/ts/src/cbor.ts
// Deterministic CBOR encoder — the Rhizomatic v0 canonicalization profile (RFC 8949 §4.2.1,
// as refined in spec/01-delta.ERRATA.md). Hand-rolled on purpose: this must reproduce the Rust
// encoder byte-for-byte, and the only way to guarantee that is to own every byte.
//
// Supported data items (exactly what the delta model needs): text string, float (numbers),
// bool, definite-length array, definite-length map (text-string keys, sorted).

export type CborValue =
  | { readonly t: "tstr"; readonly v: string }
  | { readonly t: "float"; readonly v: number }
  | { readonly t: "bool"; readonly v: boolean }
  | { readonly t: "array"; readonly v: readonly CborValue[] }
  | { readonly t: "map"; readonly v: ReadonlyArray<readonly [string, CborValue]> };

export const tstr = (v: string): CborValue => ({ t: "tstr", v });
export const float = (v: number): CborValue => ({ t: "float", v });
export const bool = (v: boolean): CborValue => ({ t: "bool", v });
export const array = (v: readonly CborValue[]): CborValue => ({ t: "array", v });
export const map = (v: ReadonlyArray<readonly [string, CborValue]>): CborValue => ({ t: "map", v });

class ByteSink {
  private readonly bytes: number[] = [];
  push(...b: number[]): void {
    for (const x of b) this.bytes.push(x & 0xff);
  }
  pushBytes(arr: Uint8Array): void {
    for (const x of arr) this.bytes.push(x);
  }
  toUint8Array(): Uint8Array {
    return Uint8Array.from(this.bytes);
  }
}

// Write a CBOR head: major type (high 3 bits) plus an unsigned argument, shortest form.
function writeHead(sink: ByteSink, major: number, arg: number): void {
  const mt = major << 5;
  if (arg < 24) {
    sink.push(mt | arg);
  } else if (arg < 0x100) {
    sink.push(mt | 24, arg);
  } else if (arg < 0x10000) {
    sink.push(mt | 25, (arg >> 8) & 0xff, arg & 0xff);
  } else if (arg <= 0xffffffff) {
    sink.push(mt | 26, (arg >>> 24) & 0xff, (arg >>> 16) & 0xff, (arg >>> 8) & 0xff, arg & 0xff);
  } else {
    // Only reachable for absurd lengths; included for totality.
    const hi = Math.floor(arg / 0x100000000);
    const lo = arg >>> 0;
    sink.push(
      mt | 27,
      (hi >>> 24) & 0xff,
      (hi >>> 16) & 0xff,
      (hi >>> 8) & 0xff,
      hi & 0xff,
      (lo >>> 24) & 0xff,
      (lo >>> 16) & 0xff,
      (lo >>> 8) & 0xff,
      lo & 0xff,
    );
  }
}

const fbuf = new DataView(new ArrayBuffer(8));

// Returns the IEEE-754 binary16 bit pattern for n if (and only if) f16 represents n exactly,
// else null. n must be finite and exactly representable as f32 (caller guarantees both).
function tryF16Bits(n: number): number | null {
  fbuf.setFloat32(0, n);
  const bits = fbuf.getUint32(0);
  const sign = ((bits >>> 31) & 1) << 15;
  const exp = (bits >>> 23) & 0xff;
  const mant = bits & 0x7fffff;
  if (exp === 0 && mant === 0) return sign; // zero (-0 already normalized away by caller)
  const e = exp - 127; // unbiased exponent (f32 subnormals land at -127 and fall through)
  if (e >= -14 && e <= 15) {
    // f16 normal range: the 23-bit mantissa must fit in 10 bits.
    if ((mant & 0x1fff) !== 0) return null;
    return sign | ((e + 15) << 10) | (mant >>> 13);
  }
  if (e >= -24 && e <= -15) {
    // f16 subnormal range: value must be an exact multiple of 2^-24.
    const shift = -(e + 1); // 14..23
    const sig = 0x800000 | mant; // full 24-bit significand
    if ((sig & ((1 << shift) - 1)) !== 0) return null;
    return sign | (sig >>> shift);
  }
  return null;
}

// ERRATA D1: numbers encode as float only, in the shortest of f16/f32/f64 that represents the
// value exactly (RFC 8949 §4.2.1). -0.0 is normalized to +0.0.
function writeFloat(sink: ByteSink, value: number): void {
  if (!Number.isFinite(value)) {
    throw new Error(`non-finite number is not representable: ${value}`);
  }
  const n = value + 0; // normalize -0 to +0
  if (Math.fround(n) === n) {
    const h = tryF16Bits(n);
    if (h !== null) {
      sink.push(0xf9, (h >>> 8) & 0xff, h & 0xff);
      return;
    }
    fbuf.setFloat32(0, n);
    sink.push(0xfa, fbuf.getUint8(0), fbuf.getUint8(1), fbuf.getUint8(2), fbuf.getUint8(3));
  } else {
    fbuf.setFloat64(0, n);
    sink.push(
      0xfb,
      fbuf.getUint8(0),
      fbuf.getUint8(1),
      fbuf.getUint8(2),
      fbuf.getUint8(3),
      fbuf.getUint8(4),
      fbuf.getUint8(5),
      fbuf.getUint8(6),
      fbuf.getUint8(7),
    );
  }
}

function cmpBytes(a: Uint8Array, b: Uint8Array): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    const d = a[i]! - b[i]!;
    if (d !== 0) return d;
  }
  return a.length - b.length;
}

const utf8 = new TextEncoder();

function encodeInto(sink: ByteSink, val: CborValue): void {
  switch (val.t) {
    case "tstr": {
      const bytes = utf8.encode(val.v.normalize("NFC"));
      writeHead(sink, 3, bytes.length);
      sink.pushBytes(bytes);
      return;
    }
    case "bool":
      sink.push(val.v ? 0xf5 : 0xf4);
      return;
    case "float":
      writeFloat(sink, val.v);
      return;
    case "array":
      writeHead(sink, 4, val.v.length);
      for (const item of val.v) encodeInto(sink, item);
      return;
    case "map": {
      // ERRATA D4: sort entries by bytewise lex order of the encoded key.
      const entries = val.v.map(([k, v]) => {
        const ks = new ByteSink();
        encodeInto(ks, tstr(k));
        return { key: ks.toUint8Array(), value: v };
      });
      entries.sort((a, b) => cmpBytes(a.key, b.key));
      for (let i = 1; i < entries.length; i++) {
        if (cmpBytes(entries[i - 1]!.key, entries[i]!.key) === 0) {
          throw new Error("duplicate map key in canonical CBOR");
        }
      }
      writeHead(sink, 5, entries.length);
      for (const e of entries) {
        sink.pushBytes(e.key);
        encodeInto(sink, e.value);
      }
      return;
    }
  }
}

export function encode(val: CborValue): Uint8Array {
  const sink = new ByteSink();
  encodeInto(sink, val);
  return sink.toUint8Array();
}

File: implementations/rust/src/cbor.rs
//! Deterministic CBOR encoder — Rhizomatic v0 profile (spec/01-delta.ERRATA.md D1-D4).
//! Must reproduce ../ts/src/cbor.ts byte-for-byte.

use unicode_normalization::UnicodeNormalization;

#[derive(Debug, Clone, PartialEq)]
pub enum CborValue {
    Tstr(String),
    Float(f64),
    Bool(bool),
    Array(Vec<CborValue>),
    /// text-string keys; sorted at encode time (ERRATA D4)
    Map(Vec<(String, CborValue)>),
}

pub fn encode(value: &CborValue) -> Vec<u8> {
    let mut out = Vec::new();
    encode_into(&mut out, value);
    out
}

/// CBOR head: major type (high 3 bits) plus unsigned argument, shortest form.
fn write_head(out: &mut Vec<u8>, major: u8, arg: u64) {
    let mt = major << 5;
    if arg < 24 {
        out.push(mt | arg as u8);
    } else if arg < 0x100 {
        out.push(mt | 24);
        out.push(arg as u8);
    } else if arg < 0x1_0000 {
        out.push(mt | 25);
        out.extend_from_slice(&(arg as u16).to_be_bytes());
    } else if arg <= 0xffff_ffff {
        out.push(mt | 26);
        out.extend_from_slice(&(arg as u32).to_be_bytes());
    } else {
        out.push(mt | 27);
        out.extend_from_slice(&arg.to_be_bytes());
    }
}

/// Returns the IEEE-754 binary16 bit pattern for `n` if (and only if) f16 represents it exactly.
/// `n` must be finite and exactly representable as f32 (caller guarantees both).
fn try_f16_bits(n: f64) -> Option<u16> {
    let bits = (n as f32).to_bits();
    let sign = (((bits >> 31) & 1) as u16) << 15;
    let exp = ((bits >> 23) & 0xff) as i32;
    let mant = bits & 0x7f_ffff;
    if exp == 0 && mant == 0 {
        return Some(sign); // zero (-0 already normalized away by caller)
    }
    let e = exp - 127; // unbiased exponent (f32 subnormals land at -127 and fall through)
    if (-14..=15).contains(&e) {
        // f16 normal range: the 23-bit mantissa must fit in 10 bits.
        if mant & 0x1fff != 0 {
            return None;
        }
        return Some(sign | (((e + 15) as u16) << 10) | (mant >> 13) as u16);
    }
    if (-24..=-15).contains(&e) {
        // f16 subnormal range: value must be an exact multiple of 2^-24.
        let shift = -(e + 1) as u32; // 14..=23
        let sig = 0x80_0000u32 | mant; // full 24-bit significand
        if sig & ((1 << shift) - 1) != 0 {
            return None;
        }
        return Some(sign | (sig >> shift) as u16);
    }
    None
}

// ERRATA D1: numbers encode as float only, in the shortest of f16/f32/f64 that represents the
// value exactly (RFC 8949 §4.2.1). -0.0 is normalized to +0.0.
fn write_float(out: &mut Vec<u8>, value: f64) {
    assert!(
        value.is_finite(),
        "non-finite number is not representable: {value}"
    );
    let n = value + 0.0; // normalize -0 to +0
    if (n as f32) as f64 == n {
        if let Some(h) = try_f16_bits(n) {
            out.push(0xf9);
            out.extend_from_slice(&h.to_be_bytes());
            return;
        }
        out.push(0xfa);
        out.extend_from_slice(&(n as f32).to_be_bytes());
    } else {
        out.push(0xfb);
        out.extend_from_slice(&n.to_be_bytes());
    }
}

fn encode_into(out: &mut Vec<u8>, value: &CborValue) {
    match value {
        CborValue::Tstr(s) => {
            let normalized: String = s.nfc().collect();
            let bytes = normalized.as_bytes();
            write_head(out, 3, bytes.len() as u64);
            out.extend_from_slice(bytes);
        }
        CborValue::Bool(b) => out.push(if *b { 0xf5 } else { 0xf4 }),
        CborValue::Float(n) => write_float(out, *n),
        CborValue::Array(items) => {
            write_head(out, 4, items.len() as u64);
            for item in items {
                encode_into(out, item);
            }
        }
        CborValue::Map(entries) => {
            // ERRATA D4: sort entries by bytewise lex order of the encoded key.
            let mut encoded: Vec<(Vec<u8>, &CborValue)> = entries
                .iter()
                .map(|(k, v)| {
                    let mut kb = Vec::new();
                    encode_into(&mut kb, &CborValue::Tstr(k.clone()));
                    (kb, v)
                })
                .collect();
            encoded.sort_by(|a, b| a.0.cmp(&b.0));
            for w in encoded.windows(2) {
                assert!(w[0].0 != w[1].0, "duplicate map key in canonical CBOR");
            }
            write_head(out, 5, encoded.len() as u64);
            for (kb, v) in encoded {
                out.extend_from_slice(&kb);
                encode_into(out, v);
            }
        }
    }
}
